### ISSUE TITLE
Support non-null assertion for destructuring

### ISSUE THREAD

**Issue by vilic** [Microsoft/TypeScript#17390]
Currently if we want to retrieve a value nested inside a nullable property using destructuring, we have to split a single destructuring statement into two even if we know that the nullable property would always be non-null in under that branch.

For example:

```ts
interface Foo {
  bar?: {
    bia: number;
    pia?: number;
  };
}

let foo: Foo = ...;

// Currently we have to write something like this:
let {bar} = foo;
let {bia, pia} = bar!;

// But ideally we should be able to write something like:
let {
  bar!: {
    bia,
    pia,
  },
} = foo;

// Or even:
let {
  bar!: {
    bia,
    pia!, // <--
  },
} = foo;
```

**Comment by Shlomibo** [comment]
Could be achieved with (or with better typings):

``` typescript
ley {
  bar: {
    bia,
    pia = 0,
  } = {} as Record<string, any> // Or Partial<Bar> if we have a Bar type,
} = foo;
```
What's bothering me, is that we have to have an explicit type assertion on the default value in order to being able to destruct it. It would be better if the default-value type would be inferred to be Partial, or other - destructuring supporting - type.

**Comment by RyanCavanaugh** [comment]
General opinion from the design meeting was that the destructuring syntax is already overly complex and hard to follow, so we'd prefer to not add any new syntax there unless it's absolutely necessary.

**Comment by isaiahtaylor** [comment]
Hey guys, it's been a few years on this one. I understand the original opinion that the destructuring syntax is complex. However, I believe the syntax is not hardened and solidified such that it's not going anywhere and is best practice. I think TypeScript should affirm that best practice with type checking. As it is, we resort to much more complex patterns due to the lack of this feature.


```ts
interface Office {
  company: {
    name : string;
    industry?: string;
    address?: {
      street: string;
      city: string;
    }
  }
}


// How it should be

function doOffice(office: Office) {
  const { company: { name, industry, address?: { street, city } } } = office;

  ...

  // explicit null checks
  if (!street || !city) {
    //do something
  }
}


// How it is now

function doOffice(office: Office) {
  const { company: { name, industry, address } } = office;

  if (!address) {
    //do something
  }

  const { street, city } = address;
}

```

As objects grow in complexity, this becomes tedious and very hard to read. Thanks for considering TS team!

**Comment by IARI** [comment]
Please reconsider.
writing something simple as
```
const { user } = auth
```
becomes useless, if i have to check it for null, or need an extra line.
imho that defeatsx the purpose of the simple use case for destructuring.
